execute/plugin.go

```diff
@@ -537,9 +544,73 @@ func (p *Plugin) validateReport(
 		return false, cciptypes.ExecutePluginReport{}, nil
 	}
 
+	// check that the messages in the report are not already executed onchain.
+	// note that this involves a set of DB queries, hence why its last in the checks.
+	seqNrRangesBySource := getSeqNrRangeBySource(decodedReport.ChainReports)
+	err = p.checkAlreadyExecuted(ctx, lggr, seqNrRangesBySource)
+	if errors.Is(err, errAlreadyExecuted) {
+		// Some messages in the report have already
+		// been executed, so we don't want to re-execute them.
+		// This gives the exec plugin a chance to remedy the situation
+		// by selecting a different set of messages.
+		return false, decodedReport, nil
+	}
+	if err != nil {
+		// TODO: should we return true here if we couldn't check for already executed messages?
+		return false, decodedReport, fmt.Errorf("checking for already executed messages failed: %w", err)
+	}
+
 	return true, decodedReport, nil
 }
 
+// checkAlreadyExecuted checks if the messages in the report have already been executed
+// on the destination chain. It queries the DB for executed messages in the given sequence
+// number range for each source chain in the report.
+func (p *Plugin) checkAlreadyExecuted(
+	ctx context.Context,
+	lggr logger.Logger,
+	seqNrRangesBySource map[cciptypes.ChainSelector]cciptypes.SeqNumRange,
+) error {
+	// TODO: batch these queries? these are all DB reads.
+	// maybe some alternative queries exist.
+	for sourceChainSelector, seqNrRange := range seqNrRangesBySource {
+		executed, err := p.ccipReader.ExecutedMessages(ctx, sourceChainSelector, seqNrRange)
+		if err != nil {
+			return fmt.Errorf("couldn't check if messages already executed: %w", err)
+		}
+
+		if len(executed) > 0 {
+			// Some subset of messages have been executed, we can't accept this report.
```

Comment:
An alternative to this "all or nothing" logic here is maybe we can have a % of mgs in the report that have been executed, if > 50% or something, then don't transmit, if < 50%, then still worth transmitting. Will leave this kinda logic for another PR, I feel like it'd be something good to test in the high-traffic envs like load tests.